File: src/test/java/application/game/cancel/CancelGameControllerTest.java
package application.game.cancel;

import application.AbstractApplicationTest;
import com.matag.admin.game.game.GameRepository;
import com.matag.admin.game.join.JoinGameRequest;
import com.matag.admin.game.join.JoinGameResponse;
import com.matag.admin.game.session.GameSession;
import com.matag.admin.game.session.GameSessionRepository;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import static application.TestUtils.*;
import static com.matag.admin.game.game.GameResultType.R2;
import static com.matag.admin.game.game.GameStatusType.FINISHED;
import static com.matag.admin.game.game.GameType.UNLIMITED;
import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.http.HttpStatus.OK;

public class CancelGameControllerTest extends AbstractApplicationTest {
  @Autowired
  private GameRepository gameRepository;

  @Autowired
  private GameSessionRepository gameSessionRepository;

  @Test
  public void shouldCancelAGameWhereNobodyJoined() {
    // Given
    loginUser(USER_1_SESSION_TOKEN, USER_1_USERNAME);
    var request = JoinGameRequest.builder()
      .gameType(UNLIMITED)
      .playerOptions("player1 options")
      .build();
    var joinGameResponse = postForEntity("/game", request, JoinGameResponse.class, USER_1_SESSION_TOKEN);
    var gameId = joinGameResponse.getResponseBody().getGameId();

    // When
    var response = delete("/game/" + gameId, USER_1_SESSION_TOKEN);

    // Then
    assertThat(response.getStatus()).isEqualTo(OK);
    assertThat(gameSessionRepository.findAll()).hasSize(0);
    assertThat(gameRepository.findAll()).hasSize(0);
  }

  @Test
  public void shouldCancelAGameWhereSomebodyJoined() {
    // Given
    loginUser(USER_1_SESSION_TOKEN, USER_1_USERNAME);
    var request1 = JoinGameRequest.builder()
      .gameType(UNLIMITED)
      .playerOptions("player1 options")
      .build();
    var joinGameResponse = postForEntity("/game", request1, JoinGameResponse.class, USER_1_SESSION_TOKEN);
    var gameId = joinGameResponse.getResponseBody().getGameId();

    loginUser(USER_2_SESSION_TOKEN, USER_2_USERNAME);
    var request2 = JoinGameRequest.builder()
      .gameType(UNLIMITED)
      .playerOptions("player2 options")
      .build();
    postForEntity("/game", request2, JoinGameResponse.class, USER_2_SESSION_TOKEN);

    // When
    loginUser(USER_1_SESSION_TOKEN, USER_1_USERNAME);
    var response = delete("/game/" + gameId, USER_1_SESSION_TOKEN);

    // Then
    assertThat(response.getStatus()).isEqualTo(OK);
    var game = gameRepository.findById(gameId);
    assertThat(game).isPresent();
    assertThat(game.get().getStatus()).isEqualTo(FINISHED);
    assertThat(game.get().getResult()).isEqualTo(R2);
    assertThat(game.get().getFinishedAt()).isNotNull();

    List<GameSession> gameSessions = StreamSupport.stream(gameSessionRepository.findAll().spliterator(), false).collect(Collectors.toList());
    assertThat(gameSessions).hasSize(2);
    assertThat(gameSessions.get(0).getSession()).isNull();
    assertThat(gameSessions.get(1).getSession()).isNull();
  }
}

File: src/test/java/application/game/cancel/CancelGameControllerTest.kt
package application.game.cancel

import application.AbstractApplicationTest
import application.TestUtils
import com.matag.admin.game.game.*
import com.matag.admin.game.join.JoinGameRequest
import com.matag.admin.game.join.JoinGameResponse
import com.matag.admin.game.session.GameSession
import com.matag.admin.session.MatagSession
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.http.HttpStatus
import java.util.stream.Collectors
import java.util.stream.StreamSupport

class CancelGameControllerTest : AbstractApplicationTest() {
    @Test
    fun shouldCancelAGameWhereNobodyJoined() {
        // Given
        loginUser(TestUtils.USER_1_SESSION_TOKEN, TestUtils.USER_1_USERNAME)
        val request = JoinGameRequest(
            gameType = GameType.UNLIMITED,
            playerOptions = "player1 options",
        )
        val joinGameResponse = postForEntity(
            "/game",
            request,
            JoinGameResponse::class.java,
            TestUtils.USER_1_SESSION_TOKEN
        )
        val gameId = joinGameResponse.getResponseBody()?.gameId

        // When
        val response = delete("/game/$gameId", TestUtils.USER_1_SESSION_TOKEN)

        // Then
        assertThat(response.status).isEqualTo(HttpStatus.OK)
        assertThat(gameSessionRepository.findAll()).hasSize(0)
        assertThat(gameRepository.findAll()).hasSize(0)
    }

    @Test
    fun shouldCancelAGameWhereSomebodyJoined() {
        // Given
        loginUser(TestUtils.USER_1_SESSION_TOKEN, TestUtils.USER_1_USERNAME)
        val request1 = JoinGameRequest(
            gameType = GameType.UNLIMITED,
            playerOptions = "player1 options"
        )
        val joinGameResponse = postForEntity(
            "/game",
            request1,
            JoinGameResponse::class.java,
            TestUtils.USER_1_SESSION_TOKEN
        )
        val gameId = requireNotNull(joinGameResponse.getResponseBody()?.gameId)

        loginUser(TestUtils.USER_2_SESSION_TOKEN, TestUtils.USER_2_USERNAME)
        val request2 = JoinGameRequest(
            gameType = GameType.UNLIMITED,
            playerOptions = "player2 options"
        )
        postForEntity(
            "/game",
            request2,
            JoinGameResponse::class.java,
            TestUtils.USER_2_SESSION_TOKEN
        )

        // When
        loginUser(TestUtils.USER_1_SESSION_TOKEN, TestUtils.USER_1_USERNAME)
        val response = delete("/game/$gameId", TestUtils.USER_1_SESSION_TOKEN)

        // Then
        assertThat(response.status).isEqualTo(HttpStatus.OK)
        val game = gameRepository.findById(gameId)
        assertThat(game).isPresent()
        assertThat(game.get().status).isEqualTo(GameStatus.FINISHED)
        assertThat(game.get().result).isEqualTo(GameResultType.R2)
        assertThat(game.get().finishedAt).isNotNull()

        val gameSessions = StreamSupport.stream<GameSession?>(gameSessionRepository.findAll().spliterator(), false)
            .collect(Collectors.toList())
        assertThat<GameSession?>(gameSessions).hasSize(2)
        assertThat<MatagSession?>(gameSessions[0].session).isNull()
        assertThat<MatagSession?>(gameSessions[1].session).isNull()
    }
}